### PR TITLE
Add org and membership models

### PR DESCRIPTION
**Pull request by indirect**
As an extremely baseline step, add models that will eventually be used to track organizations and user memberships in those organizations. Includes migrations, models, relationships, validations, tests, and avo resource for orgs.

### CODE REVIEW

**Review by martinemde — APPROVED** [review]

File: app/models/org.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class Org < ApplicationRecord
```
Comment (colby-swandale) [inline app/models/org.rb:1]: I remember the organisation vs organization argument for using `Org`, but it's hard to move past this feeling that using `Org` is a bit lazy here, where everything else in the repo has their full labels.

How about we just pull the trigger on `Organization`. Thoughts @indirect @martinemde ?

Comment (martinemde) [reply app/models/org.rb:1]: We could change it if you think it's ambiguous, but it's more than just spelling. Typing less thousands of times over the life of the model is worth it. There's plenty of examples in this repo and related ones where we used a short word when a longer one was available.

I suggest we use the shortest unambiguous name in the code.

...or there's always the slightly longer "Orga" (my daughter's word for "Orchid") 😆

Comment (jenshenny) [reply app/models/org.rb:1]: I would prefer `Organization` over `Org` because if we keep the term consistent to `Organization`, then there will be one consistent naming of the term in the repo. There won't be any ambiguity between referring to the concept as organisation / organization / org.

File: app/avo/resources/membership_resource.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class MembershipResource < Avo::BaseResource
+  self.title = :id
+  self.includes = []
+  # self.search_query = -> do
+  #   scope.ransack(id_eq: params[:q], m: "or").result(distinct: false)
+  # end
+
+  field :id, as: :id
+  # Fields generated from the model
+  field :user, as: :text
```
Comment (segiddins) [inline app/avo/resources/membership_resource.rb:10]: These should be belongs_to ?

File: app/models/membership.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class Membership < ApplicationRecord
+  belongs_to :user
+  belongs_to :org
+
+  scope :confirmed, -> { where.not(confirmed_at: nil) }
```
Comment (segiddins) [inline app/models/membership.rb:5]: nit: add matching `unconfirmed` scope?

File: app/avo/resources/membership_resource.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class MembershipResource < Avo::BaseResource
+  self.title = :id
+  self.includes = []
+  # self.search_query = -> do
+  #   scope.ransack(id_eq: params[:q], m: "or").result(distinct: false)
+  # end
```
Comment (segiddins) [inline app/avo/resources/membership_resource.rb:6]: ScopeBooleanFilter for confirmed/unconfirmed

File: app/models/org.rb (hunk 1 above)
Comment (segiddins) [reply app/models/org.rb:1]: 👍🏻 to organization from me

File: app/models/org.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  validates :handle, presence: true,
+    uniqueness: { case_sensitive: false },
+    length: { within: 2..40 },
+    format: { with: /\A[A-Za-z][A-Za-z_\-0-9]*\z/ }
```
Comment (segiddins) [inline app/models/org.rb:5]: should we disallow ending on a `[_-]`?

File: app/models/org.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  validates :name, presence: true, length: { within: 2..255 }
+  validate :unique_with_user_handle
+
+  def unique_with_user_handle
+    errors.add(:handle, "has already been taken") if handle && User.where("handle = lower(?)", handle.downcase).any?
```
Comment (segiddins) [inline app/models/org.rb:10]: `User` will need a matching validation added

File: app/models/org.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  has_many :rubygems, through: :ownerships
```
Comment (segiddins) [inline app/models/org.rb:13]: missing `ownerships` association?

File: app/models/org.rb (hunk 1 above)
Comment (segiddins) [inline app/models/org.rb:1]: i think this should also include the security event module, and we should start off by emitting all the relevant security events rather than adding them in follow-ups

File: config/locales/en.yml [hunk 8]
```diff
@@ -91,6 +91,10 @@ en:
             user_id:
               already_confirmed: "is already an owner of this gem"
               already_invited: "is already invited to this gem"
+        user:
```
Comment (segiddins) [inline config/locales/en.yml:94]: will need to run fill locales

File: db/migrate/20240630025625_create_orgs.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class CreateOrgs < ActiveRecord::Migration[7.1]
+  def change
+    create_table :orgs do |t|
+      t.string :handle
```
Comment (segiddins) [inline db/migrate/20240630025625_create_orgs.rb:4]: unique index on `lower(handle)`

Comment (segiddins) [reply db/migrate/20240630025625_create_orgs.rb:4]: set lengths for the string columns?

File: db/migrate/20240630025804_create_memberships.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateMemberships < ActiveRecord::Migration[7.1]
+  def change
+    create_table :memberships do |t|
+      t.belongs_to :user, null: false, foreign_key: true
```
Comment (segiddins) [inline db/migrate/20240630025804_create_memberships.rb:4]: User needs to have a matching association declared

File: db/migrate/20240630025804_create_memberships.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+      t.belongs_to :org, null: false, foreign_key: true
+      t.timestamp :confirmed_at, default: nil
+
```
Comment (segiddins) [inline db/migrate/20240630025804_create_memberships.rb:7]: unique index on `(user_id, organization_id)`

File: app/models/org.rb (hunk 5 above)
Comment (indirect) [reply app/models/org.rb:5]: I don't think so, unless we also change the User handle rules, since we need them to be consistent.

File: app/models/org.rb (hunk 1 above)
Comment (indirect) [reply app/models/org.rb:1]: Copying events from user, the only applicable event seems to be "created", so I have added that.

File: db/migrate/20240630025625_create_orgs.rb (hunk 9 above)
Comment (indirect) [reply db/migrate/20240630025625_create_orgs.rb:4]: Since the handle column needs to work exactly the same as on User, I think we can only set lengths if we also change the rules for user handles.

File: app/models/org.rb (hunk 1 above)
Comment (segiddins) [reply app/models/org.rb:1]: also memberships added/confirmed/removed

File: Gemfile [hunk 12]
```diff
@@ -10,14 +10,14 @@ gem "aws-sdk-sqs", "~> 1.80"
...
 gem "dalli", "~> 3.2"
-gem "datadog", "~> 2.1", require: "datadog/auto_instrument"
+gem "datadog", "~> 2.1"
 gem "dogstatsd-ruby", "~> 5.5"
 gem "google-protobuf", "~> 4.27"
 gem "faraday", "~> 2.9"
 gem "faraday-retry", "~> 2.2"
 gem "good_job", "~> 3.29"
 gem "gravtastic", "~> 3.2"
-gem "honeybadger", "~> 5.5.1" # see https://github.com/rubygems/rubygems.org/pull/4598
```
Comment (segiddins) [inline Gemfile:20]: why this change?

File: app/models/org.rb (hunk 5 above)
Comment (segiddins) [reply app/models/org.rb:7]: in that case, let's extract the pattern to a constant under `Patterns` and use that for both

File: db/migrate/20240630025625_create_orgs.rb (hunk 9 above)
Comment (segiddins) [reply db/migrate/20240630025625_create_orgs.rb:4]: there's a length validation for `handle` in the model

File: Gemfile (hunk 12 above)
Comment (indirect) [reply Gemfile:20]: requiring the gem in the gemfile was too early to configure eg logging. moving the require to the initializer means we can suppress the warning in development about exceptions not being reported to the web service.

File: app/models/org.rb (hunk 7 above)
Comment (indirect) [reply app/models/org.rb:15]: Whoops, meant to remove this for now. Making `ownerships` polymorphic and adding them to orgs is a future PR.

File: db/migrate/20240630025625_create_orgs.rb (hunk 9 above)
Comment (indirect) [reply db/migrate/20240630025625_create_orgs.rb:4]: user full_name and handle columns don't have db limits and I was just copying ¯\\_(ツ)\_/¯ added limits

File: app/models/org.rb (hunk 1 above)
Comment (colby-swandale) [reply app/models/org.rb:1]: @indirect saw this has been marked as resolved, but the code hasn't changed in regards to this. Did you end up making a decision?

Comment (indirect) [reply app/models/org.rb]: @colby-swandale whoops, sorry about that. I don't personally think the long version is worth it, but I'm fine with the majority team consensus to go with `Organization`. I just pushed a commit that renames everything, right before I saw your comment. 😅

File: config/locales/en.yml [hunk 13]
```diff
@@ -89,11 +89,19 @@ en:
           attributes:
             expires_at:
               inclusion: "must be in the future"
+        org:
```
Comment (jenshenny) [inline config/locales/en.yml:92]: This needs to be updated because of the rename to organization.

```suggestion
        organization:
```

**Review by jenshenny — APPROVED** [review]
🎉

File: app/models/user.rb [hunk 14]
```diff
@@ -358,4 +358,8 @@ def record_email_verified_event
   def record_password_update_event
     record_event!(Events::UserEvent::PASSWORD_CHANGED)
   end
+
+  def unique_with_org_handle
+    errors.add(:handle, "has already been taken") if handle && Organization.where("handle = lower(?)", handle.downcase).any?
```
Comment (segiddins) [inline app/models/user.rb:363]:
```suggestion
    errors.add(:handle, "has already been taken") if handle && Organization.where("lower(handle) = ?", handle.downcase).any?
```

is this what you meant?

**Review by segiddins — APPROVED** [review]

File: app/models/user.rb (hunk 14 above)
Comment (indirect) [reply app/models/user.rb:363]: whoops, yes 🤦🏻

File: config/locales/en.yml (hunk 13 above)
Comment (indirect) [reply config/locales/en.yml]: thank you!
